File: app/src/main/java/starbright/com/projectegg/features/ingredients/CartBottomSheetDialogFragment.java
package starbright.com.projectegg.features.ingredients;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetDialogFragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.annotations.NonNull;
import starbright.com.projectegg.R;
import starbright.com.projectegg.data.local.model.Ingredient;

public class CartBottomSheetDialogFragment extends BottomSheetDialogFragment
        implements IngredientsCartAdapter.Listener {

    public static final String EXTRA_EVENT_CART = "EXTRA_EVENT_CART";

    private static final int SPAN_COUNT = 4;

    @BindView(R.id.tv_total_ingredient)
    TextView tvTotalIngredient;

    @BindView(R.id.tv_empty)
    TextView tvEmpty;

    @BindView(R.id.rv_ingredients_cart)
    RecyclerView rvIngredientsCart;

    private SheetListener mSheetListener;
    private IngredientsCartAdapter mAdapter;
    private List<Ingredient> mCart;

    public CartBottomSheetDialogFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.partial_bottom_sheet, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view,
                              @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        updateCartCountView();
        setupRvIngredientCart();
    }

    @Override
    public void onClearItemClickedListener(int position) {
        updateView(position);
        mSheetListener.updateCart(mCart);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mSheetListener = null;
    }

    private void setupRvIngredientCart() {
        rvIngredientsCart.setLayoutManager(new GridLayoutManager(getActivity(), SPAN_COUNT,
                LinearLayoutManager.VERTICAL, false));
        mAdapter = new IngredientsCartAdapter(getActivity(), mCart);
        mAdapter.setListener(this);
        rvIngredientsCart.setAdapter(mAdapter);
        tvEmpty.setVisibility(mCart.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private void updateView(int position) {
        mCart.remove(position);
        updateCartCountView();
        updateList();
    }

    private void updateList() {
        mAdapter.notifyDataSetChanged();
        tvEmpty.setVisibility(mCart.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private void updateCartCountView() {
        tvTotalIngredient.setText(String.valueOf(mCart.size()));
    }

    public void setCartIngredient(List<Ingredient> cart) {
        mCart = cart;
    }

    public void setListener(SheetListener listener) {
        mSheetListener = listener;
    }

    @OnClick(R.id.btn_submit)
    void onClickedSubmit() {
        mSheetListener.submitButtonClicked();
    }

    interface SheetListener {
        void updateCart(List<Ingredient> ingredients);

        void submitButtonClicked();
    }
}

File: app/src/main/java/starbright/com/projectegg/features/ingredients/CartBottomSheetDialogFragment.kt
/*
 * Copyright (c) by Andreas (oentoro.andreas@gmail.com)
 * created at 25 - 7 - 2020.
 */

package starbright.com.projectegg.features.ingredients

import starbright.com.projectegg.databinding.PartialBottomSheetBinding
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import io.reactivex.annotations.NonNull
import starbright.com.projectegg.R
import starbright.com.projectegg.data.model.Ingredient

private const val ITEM_WIDTH = 120F
class CartBottomSheetDialogFragment : BottomSheetDialogFragment() {

    private var _binding: PartialBottomSheetBinding? = null
    private val binding get() = _binding!!


    private lateinit var adapter: IngredientsCartAdapter
    var cart: MutableList<Ingredient> = mutableListOf()
    var sheetListener: SheetListener? = null

    override fun onCreateView(@NonNull inflater: LayoutInflater,
                              @NonNull container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _binding = PartialBottomSheetBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(@NonNull view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnSubmit.setOnClickListener {
            sheetListener?.submitButtonClicked()
        }
        setupRvIngredientCart()
        updateButtonView()
    }

    override fun onDetach() {
        super.onDetach()
        sheetListener = null
    }

    private fun setupRvIngredientCart() {
        activity?.let {
            adapter = IngredientsCartAdapter(it, cart).also {
                it.listener = { selectedPosition ->
                    updateView(selectedPosition)
                    sheetListener?.onItemRemovedFromCart()
                }
            }
        }

        binding.rvIngredientsCart.let {
            it.layoutManager = GridLayoutManager(activity, calculateColumns())
            it.adapter = adapter
        }

        binding.tvEmpty.visibility = if (cart.isEmpty()) View.VISIBLE else View.GONE
    }

    private fun calculateColumns(): Int {
        val displayMetrics: DisplayMetrics = context?.resources?.displayMetrics ?: DisplayMetrics()
        val screenWidthDp = displayMetrics.widthPixels / displayMetrics.density
        return (screenWidthDp / ITEM_WIDTH + 0.5).toInt()
    }

    private fun updateView(position: Int) {
        cart.removeAt(position)
        updateList()
        updateButtonView()
    }

    private fun updateList() {
        adapter.notifyDataSetChanged()
        binding.tvEmpty.visibility = if (cart.isEmpty()) View.VISIBLE else View.GONE
    }

    private fun updateButtonView() {
        binding.btnSubmit.isEnabled = cart.isNotEmpty()
    }

    interface SheetListener {
        fun onItemRemovedFromCart()
        fun submitButtonClicked()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
